dali/imgcodec/tools/imagemagick_test.cc

```diff
@@ -0,0 +1,283 @@
+file is processed and the filters do not apply.
+
+The following OPTIONS are available:
+--help   -h           Show this help message
+--filter=... -r=...   Regex to filter the files with, for example '.*\\.jpeg'
+--print  -p           Only print ImageMagick's output and don't compare it with
+--batch=N  -b=N       Runs ImageMagick on batches of N images (default is 1024)
+--identify=... -i=... Path of `identify` tool, default is /usr/bin/identify
+--jobs=N  -j=N        Number of concurrent jobs to run, defaults to number of cores
+)";
+
+
+namespace dali {
+namespace imgcodec {
+namespace test {
+
+void fail(std::ostream &stream, const std::string &filename, const std::string &message) {
+  stream << "FAIL" << "\t" << filename << ": " << message << std::endl;
+}
+
+void ok(std::ostream &stream, const std::string &filename) {
+  stream << "OK" << "\t" << filename << std::endl;
+}
+
+class ImgcodecTester {
+ public:
+  ImgcodecTester() {
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("jpeg", std::make_shared<JpegParser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("png", std::make_shared<PngParser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("bmp", std::make_shared<BmpParser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("tiff", std::make_shared<TiffParser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("pnm", std::make_shared<PnmParser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("jpeg2000", std::make_shared<Jpeg2000Parser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("webp", std::make_shared<WebpParser>()));
+  }
+
+  std::optional<TensorShape<>> shape_of(const std::string &filename) const {
+    auto img = ImageSource::FromFilename(filename);
+    auto fmt = this->format_registry_.GetImageFormat(&img);
+    if (fmt == nullptr) return {};
+    auto image_info = fmt->Parser()->GetInfo(&img);
+    return image_info.shape;
+  }
+
+  ImageFormatRegistry format_registry_;
+};
+
+struct Env {
+  std::string identify_path;
+  std::string directory;
+  std::regex filter;
+  bool print;
+  unsigned batch;
+  unsigned jobs;
+  ImgcodecTester imgcodec_tester = {};
+  std::mutex mutex = {};
+};
+
+std::vector<std::string> get_batch(const Env &env,
+                                   std::filesystem::recursive_directory_iterator &it) {
+  std::vector<std::string> filenames;
+  filenames.reserve(env.batch);
+  while (it != std::filesystem::end(it) && filenames.size() < env.batch) {
+    const auto& entry = *(it++);
+    if (entry.is_regular_file()) {
+      const auto path = entry.path().string();
+      if (std::regex_match(path, env.filter))
+        filenames.push_back(path);
+    }
+  }
+  return filenames;
+}
+
+std::optional<TensorShape<>> scan_imagemagick_shape(FILE *pipe) {
```

Comment:
std::optional is C++17 and for now we need to compile DALI with C++14.

Perhaps we can just return `TensorShape<>{}` when the shape can't be extracted and then detect the error case with `shape.empty()`

Reply:
Don't we support C++17 in cpu-only code already? I recall @mzient adding such option.

Reply:
Sorry, I thought that C++17 is allowed in host-only code